File: benchmark/operand/operand_bench.py
import numpy as np, time, sys
# operand_bench.py — NumPy twin of operand_bench.cs (identical keys).
# Layout classes the op×layout×dtype matrix omits: 1-D contig/strided/reversed,
# scalar operand, mixed operand layouts (C+F, C+T), binary broadcast (row+col),
# column-broadcast unary.


def best(f, it, wm, rd):
    for _ in range(wm):
        f()
    b = float("inf")
    for _ in range(rd):
        t = time.perf_counter()
        for _ in range(it):
            f()
        b = min(b, (time.perf_counter() - t) / it)
    return b * 1000.0


N1, R, C = 1_000_000, 1000, 1000
it, wm, rd = 30, 6, 3
DTYPES = [("f64", np.float64), ("f32", np.float32), ("f16", np.float16),
          ("i32", np.int32), ("i64", np.int64), ("c128", np.complex128)]

out = []
for dn, dt in DTYPES:
    a1 = ((np.arange(N1) % 17) + 1).astype(dt)
    a1s = a1[::2]
    a1r = a1[::-1]
    a2 = ((np.arange(R * C) % 17) + 1).astype(dt).reshape(R, C)
    a2F = np.asfortranarray(a2)
    a2T = a2.T
    row = a2[0:1, :]
    col = a2[:, 0:1]
    colb = np.broadcast_to(col, (R, C))
    sc = dt(2)
    cases = {
        "1d_C": lambda: a1 + a1,
        "1d_strided": lambda: a1s + a1s,
        "1d_rev": lambda: a1r + a1r,
        "scalar_rhs": lambda: a2 + sc,
        "scalar_lhs": lambda: sc + a2,
        "mix_C_F": lambda: a2 + a2F,
        "mix_C_T": lambda: a2 + a2T,
        "bcast_row": lambda: a2 + row,
        "bcast_col": lambda: a2 + col,
        "colbcast_unary": lambda: np.positive(colb),
    }
    for k, fn in cases.items():
        try:
            fn()
            out.append(f"{k}|{dn}\t{best(fn, it, wm, rd):.6g}")
        except Exception as e:
            sys.stderr.write(f"{k}|{dn}: {type(e).__name__}: {e}\n")

print("\n".join(out))
sys.stderr.write(f"[opnd_layout_bench.py] {len(out)} rows; numpy {np.__version__}\n")

File: benchmark/operand/operand_bench.cs
#:project K:/source/NumSharp/src/NumSharp.Core/NumSharp.Core.csproj
#:property AssemblyName=NumSharp.DotNetRunScript
#:property PublishAot=false
#:property AllowUnsafeBlocks=true
// =============================================================================
// operand_bench.cs — NumSharp side of the benchmark/operand subsystem: the layout
// classes the per-operand layout grid (benchmark/layout) can't express. Companion:
// operand_bench.py (identical keys). Driven + rendered by operand_sheet.py.
//   1-D contiguous/strided/reversed · scalar operand (lhs/rhs) · mixed operand
//   layouts (C+F, C+T) · binary broadcast (row + col) · column-broadcast unary.
// Key: {case}|{dt}\t{ms}.  Run: dotnet run -c Release - < benchmark/operand/operand_bench.cs
// =============================================================================
using System.Diagnostics;
using NumSharp;

var dbg = Attribute.GetCustomAttribute(typeof(np).Assembly, typeof(DebuggableAttribute)) as DebuggableAttribute;
if (dbg?.IsJITOptimizerDisabled ?? false) { Console.Error.WriteLine("FATAL: Debug core — use dotnet run -c Release - < file"); return; }

double Best(Action f, int it, int wm, int rd)
{
    for (int i = 0; i < wm; i++) f();
    double b = double.MaxValue;
    for (int r = 0; r < rd; r++) { var sw = Stopwatch.StartNew(); for (int i = 0; i < it; i++) f(); b = Math.Min(b, sw.Elapsed.TotalMilliseconds / it); }
    return b;
}
void Row(string id, double ms) => Console.WriteLine($"{id}\t{ms:G17}");

const int N1 = 1_000_000, R = 1000, C = 1000;   // 1-D 1M ; 2-D 1M
const int it = 30, wm = 6, rd = 3;
var DTYPES = new (string name, NPTypeCode tc)[]
{
    ("f64", NPTypeCode.Double), ("f32", NPTypeCode.Single), ("f16", NPTypeCode.Half),
    ("i32", NPTypeCode.Int32), ("i64", NPTypeCode.Int64), ("c128", NPTypeCode.Complex),
};

Console.Error.WriteLine($"[operand_bench] cores={Environment.ProcessorCount}");
foreach (var (dn, tc) in DTYPES)
{
    NDArray a1, a1s, a1r, a2, a2F, a2T, row, col, colb, sc;
    try
    {
        a1 = ((np.arange(N1) % 17) + 1).astype(tc);
        a1s = a1["::2"]; a1r = a1["::-1"];
        a2 = ((np.arange(R * C) % 17) + 1).astype(tc).reshape(R, C);
        a2F = a2.copy(order: 'F');
        a2T = a2.T;                                   // (C,R) F-contig view; square so same shape
        row = a2["0:1, :"];                           // (1,C) → binary row-broadcast
        col = a2[":, 0:1"];                           // (R,1) → binary col-broadcast
        colb = np.broadcast_to(col, new Shape(R, C)); // (R,C) inner stride-0 → unary on col-broadcast
        sc = NDArray.Scalar(2, tc);                   // 0-d scalar operand
    }
    catch (Exception e) { Console.Error.WriteLine($"build {dn}: {e.GetType().Name}: {e.Message.Split('\n')[0]}"); continue; }

    void Case(string key, Action body)
    {
        try { body(); Row($"{key}|{dn}", Best(body, it, wm, rd)); }
        catch (Exception e) { Console.Error.WriteLine($"{key}|{dn}: {e.GetType().Name}: {e.Message.Split('\n')[0]}"); }
    }

    Case("1d_C",           () => { var _ = a1 + a1; });
    Case("1d_strided",     () => { var _ = a1s + a1s; });
    Case("1d_rev",         () => { var _ = a1r + a1r; });
    Case("scalar_rhs",     () => { var _ = a2 + sc; });
    Case("scalar_lhs",     () => { var _ = sc + a2; });
    Case("mix_C_F",        () => { var _ = a2 + a2F; });
    Case("mix_C_T",        () => { var _ = a2 + a2T; });
    Case("bcast_row",      () => { var _ = a2 + row; });
    Case("bcast_col",      () => { var _ = a2 + col; });
    Case("colbcast_unary", () => { var _ = np.positive(colb); });
}
Console.Error.WriteLine("[operand_bench] done");
